### PR TITLE
HTTP Basic Auth filter

### PR DESCRIPTION
**Pull request by zhaohuabing**
The implementation for [HTTP Basic Auth ](https://datatracker.ietf.org/doc/html/rfc7617)filter. 

Description: HTTP Basic Authn is a frequently requested feature from users, I'd like to have Envoy to support it.

Related issue: https://github.com/envoyproxy/envoy/issues/15365

Sponsor: [wbpcode](https://github.com/wbpcode)

Risk Level: low
Testing: Yes
Docs Changes: Yes
Release Notes: Yes
Platform Specific Features: No
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by wbpcode — COMMENTED** [review]
sounds a reasonable contribution, thanks. But I think at lease one maintainer sponsor and two reviewers two this new core extension.

> All extensions must be sponsored by an existing maintainer. Sponsorship means that the maintainer will shepherd the extension through design/code reviews. Maintainers can self-sponsor extensions if they are going to write them, shepherd them, and maintain them.

> Each extension must have two reviewers proposed for reviewing PRs to the extension. Neither of the reviewers must be a senior maintainer. Existing maintainers (including the sponsor) and other contributors can count towards this number. The initial reviewers will be codified in the [CODEOWNERS](https://github.com/envoyproxy/envoy/blob/main/CODEOWNERS) file for long term maintenance. These reviewers can be swapped out as needed.

See https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md for more detail.

If you are willing, I can help to review or even sponsor it (this extension should be simple and won't take too much bandwidth), but we still need another reviewer to ensure the quality of extension.

Or you can add this extension as `contrib` extension. `contrib` extension have a relatively loose requirement.

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:39]: It's OK to use a data format like that.

But I am not sure if it's appropriate to use the `htpasswd` as the API name.

Comment (zhaohuabing) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:39]: Thought about it again and I agreed. Which one do you like in the following, or do you have other suggestions?

* credentials
* users
* passwords

**Review by mattklein123 — COMMENTED** [review]
Seems OK to me to add if @wbpcode is willing to review.

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (wbpcode) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:39]: I am not good at that also 🤣 Maybe `users` or `passwords`, at least they are better than `htpasswd`.

Comment (zhaohuabing) [reply api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:39]: Let's go with `users` then.

**Comment by zhaohuabing** [comment]
@wbpcode Since we agreed on the API, I changed this PR to draft and proceed with implementation.

File: CODEOWNERS
Comment (wbpcode) [inline CODEOWNERS:196]: Please add sponsor (me) and initial reviewer (still me) to the owner list, then format checker wont complain. :)

**Comment by zhaohuabing** [comment]
How can I fix this?

https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/152983/logs/172
```
2023-10-19T01:49:04.5485236Z [91m/tmp/tmpo9ctbr_i/generated/rst/api-v3/extensions/filters/http/basic_auth/v3/basic_auth.proto.rst:60: WARNING: Inline substitution_reference start-string without end-string.[39;49;00m
2023-10-19T01:49:04.5486695Z Sphinx ran (2/2 cores) in 94.49305200576782s
2023-10-19T01:49:04.5487169Z BUILD FAILED
2023-10-19T01:49:04.5488423Z [31m[1mERROR: [0m/source/docs/BUILD:275:8: Executing genrule //docs:html failed: (Exit 1): bash failed: error executing command (from target //docs:html) 
2023-10-19T01:49:04.5489761Z   (cd /build/bazel_root/base/execroot/envoy && \
2023-10-19T01:49:04.5490241Z   exec env - \
2023-10-19T01:49:04.5490577Z     BAZEL_DO_NOT_DETECT_CPP_TOOLCHAIN=1 \
2023-10-19T01:49:04.5491148Z     BAZEL_FAKE_SCM_REVISION=e3b4a6e9570da15ac1caffdded17a8bebdc7dfc9 \
2023-10-19T01:49:04.5491832Z     BAZEL_LINKLIBS=-l%:libstdc++.a \
2023-10-19T01:49:04.5492292Z     BAZEL_LINKOPTS=-lm \
2023-10-19T01:49:04.5492646Z     CC=clang \
2023-10-19T01:49:04.5492954Z     CXX=clang++ \
2023-10-19T01:49:04.5493346Z     CXXFLAGS='-stdlib=libc++' \
2023-10-19T01:49:04.5494024Z     DOCS_RST_CHECK=1 \
2023-10-19T01:49:04.5494449Z     LDFLAGS='-stdlib=libc++' \
2023-10-19T01:49:04.5494914Z     PATH=/usr/sbin:/usr/bin:/sbin:/bin:/opt/llvm/bin \
2023-10-19T01:49:04.5495487Z     SPHINX_RUNNER_ARGS='-v warn' \
2023-10-19T01:49:04.5496355Z   /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; 
2023-10-19T01:49:04.5499135Z     bazel-out/k8-opt-exec-A93844C3/bin/tools/docs/sphinx_runner          ${SPHINX_RUNNER_ARGS:-}          --build_sha="${BUILD_DOCS_SHA:-}"          --version_file=./VERSION.txt          --descriptor_path=bazel-out/k8-opt-exec-A93844C3/bin/external/envoy_api/v3_proto_set.pb          bazel-out/k8-opt-exec-A93844C3/bin/docs/rst.tar.gz          bazel-out/k8-fastbuild/bin/docs/html.tar.gz
2023-10-19T01:49:04.5502037Z     ')
2023-10-19T01:49:04.5502568Z # Configuration: 59b47ef04400c56a8784305b332145459f32aba7d064706672316110f40d0c3e
2023-10-19T01:49:04.5504267Z # Execution platform: @envoy_build_tools//toolchains:rbe_linux_clang_libcxx_platform
2023-10-19T01:49:04.6110510Z Target //tools/tarball:unpack failed to build
2023-10-19T01:49:04.7111951Z [32mINFO: [0mElapsed time: 161.041s, Critical Path: 128.14s
2023-10-19T01:49:04.7113134Z [32mINFO: [0m708 processes: 653 remote cache hit, 54 internal, 1 remote.
2023-10-19T01:49:04.7193678Z [31m[1mERROR: [0mBuild failed. Not running target
2023-10-19T01:49:05.0445943Z [32mINFO: [0mStreaming build results to: https://source.cloud.google.com/results/invocations/67d01545-51a7-464d-b3e2-a3f6a31dbb72
2023-10-19T01:49:05.1933312Z ##[error]Bash exited with code '1'.
2023-10-19T01:49:05.1984296Z ##[section]Finishing: Run CI script docs
2023-10-19T01:49:05.2021023Z ##[section]Starting: Check disk space at end
```

File: api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto
Comment (wbpcode) [inline api/envoy/extensions/filters/http/basic_auth/v3/basic_auth.proto:29]:
```suggestion
// .. code-block:: yaml
//
//   users:
//     inline_string: |-
//       user1:{SHA}hashed_user1_password
//       user2:{SHA}hashed_user2_password
```

This should fix your docs error.

**Review by wbpcode — COMMENTED** [review]
Thanks. A comment to fix the docs (`rst` is powerful, but most of developers (include me) are not familiar with it orz).

This is a big PR, I can take a full through on this PR at this weekend or next Monday.

File: source/extensions/filters/http/basic_auth/basic_auth_filter.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.cc:67]:
```
absl::string_view decoded_view = decoded;
absl::string_view username = decoded_view.substr(0, colon_pos);
absl::string_view password = decoded_view.substr(colon_pos + 1);
```

File: source/extensions/filters/http/basic_auth/basic_auth_filter.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.cc:75]: Please provide a short detail string here also. For example, `error_credential_for_basic_auth`, etc.

File: source/extensions/filters/http/basic_auth/basic_auth_filter.h
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.h:39]: Couldn't we use normal asbl::flat_hash_map<std::string, std::string> directly here?

File: source/extensions/filters/http/basic_auth/basic_auth_filter.h
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.h:48]: This should be a `const` method.

File: source/extensions/filters/http/basic_auth/basic_auth_filter.h
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.h:61]: Please inherit `PassThroughDecoderFilter` rather than `PassThroughFilter ` here.

File: source/extensions/filters/http/basic_auth/config.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/config.cc:21]: `const size_t colon_pos = ...`

File: source/extensions/filters/http/basic_auth/config.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/config.cc:27]: It would be better use string view here to avoid unnecessary string copy.

File: source/extensions/filters/http/basic_auth/config.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/config.cc:31]: `name.empty()` (clang tidy will complain this. Although we haven't fixed our clang tidy ci.)


`throw EnvoyException("basic auth: invalid user name");`

File: source/extensions/filters/http/basic_auth/config.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/config.cc:33]: how about a password with format `xxxx{SHA}xxxx`? I think we should use `absl::StartWith` here to check the format of password.

File: source/extensions/filters/http/basic_auth/config.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/config.cc:37]: Please add a simple comment to explain the length checking.

And please add a similar prefix to the exception message `basic auth: ` to simplify debugging.

File: source/extensions/filters/http/basic_auth/config.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/config.cc:55]: `const std::string htpasswd = ...`

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for you great work and some comments are added.

File: source/extensions/filters/http/basic_auth/basic_auth_filter.cc
Comment (zhaohuabing) [reply source/extensions/filters/http/basic_auth/basic_auth_filter.cc:21]: It appears that static_cast can't cast between 'unsigned char *' and 'char *'. 
 ```
 static_cast from 'const std::basic_string<char>::value_type *' (aka 'const char *') to 'const unsigned char *' is not allowed
```

Comment (wbpcode) [reply source/extensions/filters/http/basic_auth/basic_auth_filter.cc:21]: got it. Then `reinterpret_cast` is acceptable here. :)

File: source/extensions/filters/http/basic_auth/basic_auth_filter.cc
Comment (zhaohuabing) [reply source/extensions/filters/http/basic_auth/basic_auth_filter.cc]: They are passed by reference, so no temporary stings will be created?

**Comment by wbpcode** [comment]
Please don't use force-update after the review has started. It will make it pretty hard to review a PR that contains lots of changes. Thanks.

File: source/extensions/filters/http/basic_auth/basic_auth_filter.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.cc:43]: nit: `return computeSHA1(password) == user->scond.hash;`

File: source/extensions/filters/http/basic_auth/basic_auth_filter.cc
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.cc:64]: Run `ci/do_ci.sh format` to fix the format.

**Comment by zhaohuabing** [comment]
> Please don't use force-update after the review has started. It will make it pretty hard to review a PR that contains lots of changes. Thanks.

Sorry for that, will not use force-update again.

File: source/extensions/filters/http/basic_auth/config.cc
Comment (zhaohuabing) [reply source/extensions/filters/http/basic_auth/config.cc]: I'm not an experienced C++ programmer, so please correct me if I'm wrong :-)

I think we need to use string instead of string_view here. We need to store the username and password in the user map, and a string_view will point to a null string because the original string is a local variable.

Comment (zhaohuabing) [reply source/extensions/filters/http/basic_auth/config.cc]: Oh, I see, you mean just use a string_view to avoid copying from the line.

File: source/extensions/filters/http/basic_auth/config.cc
Comment (zhaohuabing) [inline source/extensions/filters/http/basic_auth/config.cc:27]: Do we need to use string_view here? I think a plain simple std::string is good enough. @wbpcode

Comment (wbpcode) [reply source/extensions/filters/http/basic_auth/config.cc:27]: Both is OK for this function (because this only be called once when configuration is loaded), although string view still be better to reduce unnecessary sub-string copy/creation.

And please don't declare multiple variables by `std::string name, hash;`.

```
absl::string_view name = line_view.substr(0, colon_pos);
absl::string_view hash = line_view.substr(colon_pos + 1);

if(name.empty()) {
  throw EnvoyException("basic auth: invalid user name");
}

if(!absl::StartsWith(hash_view, "{SHA}")) {
  throw EnvoyException("basic auth: unsupported htpasswd format: please use {SHA}");
}

hash = hash.substr(5);
if (hash.length() != 28) {
  throw EnvoyException("basic auth: invalid SHA hash length");
}

users->emplace(name, User{std::string(name), std::string(hash)});

```

(Not sure whether the compiler will be happy to this code snippet, just as a nit suggestion)

**Review by wbpcode — COMMENTED** [review]
Thanks for you updates. I think this is only one last comment to be addressed.

And please add a release note for this new extension, then we can land it.

**Comment by zhaohuabing** [comment]
> Both is OK for this function (because this only be called once when configuration is loaded), although string view still be better to reduce unnecessary sub-string copy/creation.

I changed it back to std::string because the compiler was not happy about this line:

```
users->emplace(name, User{std::string(name), std::string(hash)});
```

**Comment by wbpcode** [comment]
> > Both is OK for this function (because this only be called once when configuration is loaded), although string view still be better to reduce unnecessary sub-string copy/creation.
> 
> I changed it back to std::string because the compiler was not happy about this line:
> 
> ```
> users->emplace(name, User{std::string(name), std::string(hash)});
> ```

SGTM, we any optimize it in the future anyway, if we have free time. Just merge main and add a release note. :)

**Review by wbpcode — DISMISSED** [review]
LGTM. Thanks. 

Hi, cc @mattklein123 could you also give a quick check in case I am missing something?

File: docs/root/configuration/http/http_filters/basic_auth_filter.rst
Comment (cpakulski) [inline docs/root/configuration/http/http_filters/basic_auth_filter.rst:6]: typo
s/heaer/header

File: docs/root/configuration/http/http_filters/basic_auth_filter.rst
Comment (cpakulski) [inline docs/root/configuration/http/http_filters/basic_auth_filter.rst:13]: typo
s/rqeust/request

**Review by cpakulski — COMMENTED** [review]
Few typos.
Should integration test be added to make sure that it works e-2-e?

File: docs/root/configuration/http/http_filters/basic_auth_filter.rst
Comment (wbpcode) [reply docs/root/configuration/http/http_filters/basic_auth_filter.rst:6]: weird, why our ci doesn't found it 🤔

**Comment by wbpcode** [comment]
> Should integration test be added to make sure that it works e-2-e?

SGTM. But I have no strong opinion on this because this extension is pretty simple.

**Comment by zhaohuabing** [comment]
> > Should integration test be added to make sure that it works e-2-e?
> 
> SGTM. But I have no strong opinion on this because this extension is pretty simple.

Should I add an e-2-e test? Looks like there are no e-2-e tests for HTTP filters in the test/integration directory, or I looked into the wrong place?

**Comment by wbpcode** [comment]
Integration tests are placed in sub directory of filter itself.

You can check `test/extensions/filters/http/fault/fault_filter_integration_test.cc` as a example to see how to write a integration test of http filter.

**Comment by zhaohuabing** [comment]
@wbpcode @cpakulski Integration test has been added. PTAL.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Review by mattklein123 — APPROVED** [review]
Nice!

**Pull request merged**
